### PR TITLE
[@types/stripe] - add Stripe.oauth resources

### LINKED ISSUE #40442: [@types/stripe] Stripe.oauth is not defined
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).

https://stripe.com/docs/connect/oauth-reference

### PR DESCRIPTION
**Pull request by esetnik**
Fixes #40442, #40163

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
https://stripe.com/docs/connect/oauth-reference
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/stripe/index.d.ts [hunk 1]
```diff
@@ -13755,6 +13834,13 @@ declare namespace Stripe {
             retrieve(tokenId: string, response?: IResponseFn<tokens.IToken>): Promise<tokens.IToken>;
         }
 
+        class OAuth extends StripeResource {
+            token(
```
Comment (pushplay) [inline types/stripe/index.d.ts]: Does this function support `HeaderOptions`?

File: types/stripe/index.d.ts [hunk 2]
```diff
@@ -13755,6 +13834,13 @@ declare namespace Stripe {
...
+                data: oauth.IOAuthAuthorizationCodeTokenRequest | oauth.IOAuthRefreshTokenRequest,
+                response?: IResponseFn<oauth.IOAuthToken>,
+            ): Promise<oauth.IOAuthToken>;
```
Comment (pushplay) [inline types/stripe/index.d.ts]: Can you add the `deauthorize` function as well?

File: types/stripe/index.d.ts [hunk 3]
```diff
@@ -7746,6 +7748,83 @@ declare namespace Stripe {
         }
     }
 
+    namespace oauth {
+        interface IOAuthAuthorizationCodeTokenRequest {
+            /**
+             * authorization_code when turning an authorization code into an access token
+             */
+            grant_type: 'authorization_code';
+            /**
```
Comment (pushplay) [inline types/stripe/index.d.ts]: On the new interfaces there should be a newline before block comments.

eg:
```
            /**
             * authorization_code when turning an authorization code into an access token
             */
            grant_type: 'authorization_code';

            /**
             * The value of the code
             */
            code: string;
```

**Review by pushplay — CHANGES_REQUESTED** [review]
Looks pretty good.  Thanks for copying over the property comments.  There's just a few things missing.

File: types/stripe/index.d.ts (hunk 2 above)
Comment (esetnik) [reply types/stripe/index.d.ts]: @pushplay do you prefer a full jsdoc comment? I'd like to document the parameters of deauthorize  using the same language stripe uses in their documentation.

```ts
            /**
             * When revoking access to an account, you must use an API key that matches the mode—live or test—of the authorization code (which depends on whether the client_id used was production or development).
             *
             * @param {string} client_id The client_id of the application that you'd like to disconnect the account from. The account must be connected to this application.
             * @param {string} stripe_user_id The account you'd like to disconnect from.
             * @param {HeaderOptions} options
             * @param {IResponseFn<oauth.IOAuthDeauthorizationResponse>} [response]
             * @returns {Promise<oauth.IOAuthDeauthorizationResponse>}
             * @memberof OAuth
             */
            deauthorize(
                client_id: string,
                stripe_user_id: string,
                options: HeaderOptions,
                response?: IResponseFn<oauth.IOAuthDeauthorizationResponse>,
            ): Promise<oauth.IOAuthDeauthorizationResponse>;
```

File: types/stripe/index.d.ts (hunk 3 above)
Comment (esetnik) [reply types/stripe/index.d.ts]: Maybe we should add a prettier rule for this? e.g. https://eslint.org/docs/rules/lines-around-comment

Comment (esetnik) [reply types/stripe/index.d.ts]: nvm I see you are using tslint and couldn't find any equivalent rule

File: types/stripe/index.d.ts (hunk 2 above)
Comment (esetnik) [reply types/stripe/index.d.ts]: nvm i see that it's redundant so i've only documented the `client_id` and `stripe_user_id`

File: package.json [hunk 4]
```diff
@@ -31,5 +31,6 @@
             "_comment": "This will remove husky from when we started migrating to use prettier",
             "pre-commit": "npm uninstall husky"
         }
-    }
+    },
```
Comment (pushplay) [inline package.json]: I don't know what this does to the larger project but I would revert this just to be safe.

Comment (esetnik) [reply package.json]: sorry about that. npm automatically added it when I ran `npm install`

**Comment by esetnik** [comment]
@pushplay this is marked as `Revision needed` but I don't see any feedback indicating there is an issue. Can you confirm?

**Review by pushplay — APPROVED** [review]
Thanks for the changes!  The tag `Revision needed` is the system waiting for me to take another look.

**Pull request merged**